source/containers/ArchivesList.js

```diff
@@ -20,16 +32,83 @@ function openArchive(dispatch, getState, sourceID) {
     const targetSource = archivesList.find(source => source.id === sourceID);
     // Select source
     dispatch(setSelectedSource(sourceID));
+    const isReadOnly = getSourceReadonlyStatus(sourceID);
+    dispatch(markCurrentSourceReadOnly(isReadOnly));
     // populate groups
     updateCurrentArchive();
     // navigate to archive contents
     dispatch(navigateToGroups({ groupID: "0", title: `🗂 ${targetSource.name}` }));
 }
 
+function performOfflineProcedure(dispatch, getState, sourceID, password, isOffline = false) {
+    return checkSourceHasOfflineCopy(sourceID).then(hasOffline => {
+        if (!hasOffline) {
+            return false;
+        }
+        Alert.alert(
+            isOffline ? "Offline Content (currently offline)" : "Offline Content",
+            "Would you like to try and load this archive in offline (read-only) mode?",
+            [
+                {
+                    text: "Cancel",
+                    style: "cancel",
+                    onPress: () => {
+                        dispatch(setBusyState(null));
+                    }
+                },
+                {
+                    text: "Use Offline",
+                    style: "default",
+                    onPress: () => {
+                        performSourceUnlock(dispatch, getState, sourceID, password, true)
```

Comment:
instead of calling the function and passing `dispatch` and `getState` to it, we can dispatch it (uses redux thunk):

```js
dispatch(performSourceUnlock(sourceID, password, true))
```

and when defining `performSourceUnlock` action, it can be like this:

```js
const performSourceUnlock = (sourceID, password, ...) => (dispatch, getState) => {...}
```

Reply:
Ok, but does redux thunk understand promises? Not sure how that'd work in your first example. 

`const performSourceUnlock = (sourceID, password, ...) => (dispatch, getState) => {...}`

This is still quite verbose don't you think? Having the dispatch and getState in the signature is actually shorter..

Reply:
What I mean is, `performSourceUnlock` returns a promise, and I return that so that flow continues between the offline calls. Not sure that dispatch would work in that case.